### PR TITLE
Add alltoall_ to CommTensor

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90681
* #90569
* #90565
* #90564
* __->__ #90512

This PR adds alltoall_ to the CommTensor

### CODE REVIEW

File: torch/csrc/distributed/c10d/init.cpp [hunk 1]
```diff
@@ -1477,21 +1477,21 @@ that adds a prefix to each key inserted to the store.
...
               py::arg("opts") = ::c10d::AllToAllOptions(),
               py::call_guard<py::gil_scoped_release>())
 
           .def(
               "alltoall",
               [](const c10::intrusive_ptr<::c10d::ProcessGroup>& self,
-                 std::vector<at::Tensor>& output,
-                 std::vector<at::Tensor>& input) {
+                 const std::vector<at::Tensor>& output_tensors,
+                 const std::vector<at::Tensor>& input_tensors) {
```
Comment (yifuwang) [inline torch/csrc/distributed/c10d/init.cpp:1489]: Do we expect calling this through `pg.alltoall(..)` to be backward compatible?

**Review by yifuwang — COMMENTED** [review]
Nice!

File: torch/csrc/distributed/c10d/init.cpp (hunk 1 above)
Comment (wanchaol) [reply torch/csrc/distributed/c10d/init.cpp:1489]: good point, yeah that BC breaking issue is weird, let me investigate a bit what happened there

Comment (mrshenli) [reply torch/csrc/distributed/c10d/init.cpp:1489]: The BC error is on the return type for `alltoall_`. Looks like this is recently added into BC check. This should be fine, as it is not a user-facing API, not even user-facing for PG extensions devs. 

```
2022-12-09T03:04:38.6291268Z processing existing schema:  c10d::alltoall_(Tensor[] _0, Tensor[] _1, __torch__.torch.classes.c10d.ProcessGroup _2, int _3) -> __torch__.torch.classes.c10d.Work _0
2022-12-09T03:04:38.6292120Z Can NOT find forward compatible schemas after changes for schema c10d::alltoall_(Tensor[] _0, Tensor[] _1, __torch__.torch.classes.c10d.ProcessGroup _2, int _3) -> __torch__.torch.classes.c10d.Work _0 from the following candidates:
2022-12-09T03:04:38.6292185Z [
2022-12-09T03:04:38.6292525Z c10d::alltoall_(Tensor[] _0, Tensor[] _1, __torch__.torch.classes.c10d.ProcessGroup _2, int _3) -> (Tensor[] _0, __torch__.torch.classes.c10d.Work _1)
2022-12-09T03:04:38.6292634Z ]
```

**Review by mrshenli — APPROVED** [review]
Stamp to unblock. Please address comments :)

File: torch/csrc/distributed/c10d/init.cpp [hunk 2]
```diff
@@ -1477,21 +1477,21 @@ that adds a prefix to each key inserted to the store.
...
                 return ::c10d::ops::alltoall(
-                    self, output, input, ::c10d::AllToAllOptions());
+                    self, output_tensors, input_tensors, ::c10d::AllToAllOptions());
               },
-              py::arg("output"),
-              py::arg("input"),
+              py::arg("output_tensors"),
```
[the remark this answers is not in the document]
Comment (wanchaol) [reply torch/csrc/distributed/c10d/init.cpp:1488]: Yeah make sense, updated

**Pull request closed without merging**